### PR TITLE
Use sparse checkout in release, ci and live tests

### PR DESCRIPTION
**Pull request by benbp**
This PR updates the majority of release, ci and live test pipeline jobs to use our sparse checkout template. This shaves 2-4 minutes off each job, by replacing a full git clone with one that: only grabs the root of the tree at the relevant commit, and only grabs the relevant files (service directory, all xml files, all markdown files, and in FromSource test cases, all directories of all source dependencies). For some projects with long test runs, this will not be noticeable (e.g. storage), for other projects, this amounts to almost 50% of the test run time (e.g. appconfiguration). For release, this update will remove 12-15 minutes of total time, since a lot of the release steps (pushing tags, packages, docs) have a lot of waiting time in between due to checkouts.

### CODE REVIEW

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/live.tests.yml:51]: How do we verify this doesn't break anything? I assume we need to run all test pipelines to truly verify. If we find a specific test pipeline that needs additional files how do they fix that? Should we add another template parameter that we pass through from live.tsts.yml?

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release.yml:268]: Is this going to work here? For the java version updates below they could potentially change files all over the repo.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release.yml:160]: Why do we need to pass all these extra parameters in this case? Is it because this job is mult-checkout?

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release.yml:89]: Why are the xml files needed here? Is it because the version update checks them? Is that all it checks?

Comment (benbp) [reply eng/pipelines/templates/stages/archetype-java-release.yml:89]: I have the xml files included for `set_versions.py` below. It's possible it has more minimal file dependencies in this case, but in other places I've seen it require the xml files.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-java-release.yml:160]: The default behavior when Repositories aren't specified mirrors the devops default behavior, which is to clone into the working directory `/1/s`. When multiple repositories are specified, we need to clone each to a separate path, and this is how I set up the parameters for the sparse checkout job to do that.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-java-release.yml:268]: You're right, I probably need to glob xml for `update_versions.py`. I haven't been able to get this far in release testing because of the dependency issues with `template` mentioned in the description.

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (benbp) [reply eng/pipelines/templates/jobs/live.tests.yml:51]: It's a little harder even than running all pipelines, because I worry there could be cases where we don't include certain files and tests still pass (since they'll run against found/built files). So I think the approach of a large cross-section for general pass/fail/duration parity + a small cross-section that I go over with a fine-toothed comb for differences (or script to compare test counts in the logs), is the way to go here.

I don't anticipate finding places where we require additional files, but if we do, I'd rather handle it outside of the pipeline config (i.e. via the scripts or some sort of external config), to lay the groundwork for more explicit configuration of a broad set of test file dependencies (similar to the problem we have with non-trivial test-resources.json use cases).

**Review by alzimmermsft — APPROVED** [review]
From a Java perspective this looks good

File: sdk/template/CODEOWNERS
Comment (weshaggard) [inline sdk/template/CODEOWNERS:1]: What is this file for?

File: eng/pipelines/templates/steps/checkout-source-dependencies.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: Why have all xml and md files but only some of the java files?

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release.yml:206]: What this? 

If we need the CODEOWNERS files to get the pr owners? I will say we plan to remove this from this release pipeline now that we aren't submitting PRs for the docs PRs but we just haven't cleaned this up yet.

File: sdk/template/CODEOWNERS
Comment (benbp) [reply sdk/template/CODEOWNERS:1]: The template release fails when CODEOWNERS is missing, because some of our docs scripts use it to look for that data. Apparently we haven't actually released template in a while, so I ran into this issue when trying to validate all the functionality.

File: eng/pipelines/templates/steps/checkout-source-dependencies.yml
Comment (benbp) [reply eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: We could probably narrow down the xml files to `pom*.xml`, and as well identify the specific markdown files. But they are so cheap to checkout, I figured it was simpler to glob download them, that way we don't have to do too much maintenance should the dependencies on those filetypes change slightly. We also have dependencies in some of the jobs on these filetypes that may exist outside the service directory. The java files in the repo can take up a sizable amount of checkout time (mostly mgmt libraries), so it's definitely something we want to restrict to the service directory only.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-java-release.yml:206]: Yeah there are 2 things one is the clean-up of the need for having to parse CODEOWNERS file under .github/CODEOWNERS. However my other question is why are you adding CODEOWNERS at the root as a path as well?

File: sdk/template/CODEOWNERS
Comment (weshaggard) [reply sdk/template/CODEOWNERS:1]: I think that is just a path issue we have never had CODEOWNER files directly in service directories. The only one should be under the .github folder. This file should not be needed.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-java-release.yml]: This was my mistake. Fixed to only include `.github/CODEOWNERS`.

File: sdk/template/CODEOWNERS
Comment (weshaggard) [reply sdk/template/CODEOWNERS:1]: I assume you still plan to delete this file.

File: eng/scripts/generate_from_source_pom.py
Comment (weshaggard) [inline eng/scripts/generate_from_source_pom.py:59]: Instead of this being a boolean I suggest that it be the name of the variable. It would allow opt-in and customization of the specific variable.

File: eng/pipelines/templates/steps/checkout-source-dependencies.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: I have to say this template seems a little scary to me and trying to understand it will likely be trouble in the future.

For example why do we need to add eng, "**/*.xml", "**/*.md" again here? Isn't that already included in the sparse checkout template?

File: eng/scripts/generate_from_source_pom.py
Comment (benbp) [reply eng/scripts/generate_from_source_pom.py:59]: I'm not quite sure what you mean? Do you mean not setting a type hint+default value?

File: eng/pipelines/templates/steps/checkout-source-dependencies.yml
Comment (benbp) [reply eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: I'm still testing this and have switched it to something a little simpler:

```
  - pwsh: |
      $paths = ('$(CheckoutDirectories)' | ConvertFrom-Json) -join ' '
      $command = "git sparse-checkout set eng '**/*.xml' '**/*.md' '**/test-resources.json' $paths"
      if (!$${{ parameters.CheckoutRecordings }}) {
        $command += " '!/**/session-records/*/.json'"
        $command += " '!/**/test-recordings/*/.json'"
      }
      Write-Host $command
      Invoke-Expression -Command $command
      Write-Host "Set sparse checkout paths to:"
      Get-Content .git/info/sparse-checkout
    displayName: Sparse checkout module paths
```

Comment (benbp) [reply eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: I think I can move the xml/md out of here.

Comment (benbp) [reply eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: Perhaps I should also do away with `CheckoutRecordings` and just pass in the exclusion paths from the `live.tests.yml` file.

Comment (benbp) [reply eng/pipelines/templates/steps/checkout-source-dependencies.yml:19]: Another possibility is to remove the variable passing, and instead have the pom scanner script dump out a formatted sparse checkout file that we can just drop into place? That would simplify the yaml quite a bit.

File: eng/pipelines/templates/jobs/ci.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:100]: Is there any reason these all steps next to each other? Is there an ordering issue which requires the gap?

File: eng/pipelines/templates/jobs/ci.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/ci.yml:179]: What this not working before?

Comment (benbp) [reply eng/pipelines/templates/jobs/ci.yml:179]: It works fine without specifying this parameter (as it matches the default), but since the cspell path is defined above in the sparse checkout includes, I thought it would help prevent scenarios where someone may delete/change one without the other, or wonder where the sparse checkout cspell.json include is being used.

File: eng/pipelines/templates/jobs/ci.yml
Comment (benbp) [reply eng/pipelines/templates/jobs/ci.yml:100]: Do you mean why is daily-dev-build in between these two steps? I can't recall whether this was an artifact of the order in which I edited things or not. I can't find any dependencies across steps that would prevent me from moving the dev build step lower down.

Comment (weshaggard) [reply eng/pipelines/templates/jobs/ci.yml:100]: Exactly I'd like to keep all the sparse checkout steps next to each other if possible.

File: eng/pipelines/templates/jobs/ci.yml
Comment (weshaggard) [reply eng/pipelines/templates/jobs/ci.yml:179]: might be worth a comment otherwise someone will likely remove it if they see that it matches the default value.

File: eng/scripts/generate_from_source_pom.py
Comment (weshaggard) [reply eng/scripts/generate_from_source_pom.py:59]: I'm suggesting we make this parameter a string that called pipeline_variable that we can pass in the value CheckoutDirectories in our case. If the variable is set then we set it below if not then we don't. That makes sure that we can keep the name of this variable consistent in our different usages all in the yml file and not have these as tightly coupled.

File: eng/pipelines/templates/steps/initialize-test-environment.yml
Comment (weshaggard) [inline eng/pipelines/templates/steps/initialize-test-environment.yml:56]:
```suggestion
      arguments: '--set-pipeline-variable "CheckoutDirectories" --project-list $(ProjectList)'
```
This is what I was referring to in my other comment.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (weshaggard) [inline eng/pipelines/templates/stages/archetype-java-release.yml:278]: Please have a closer look at this one I think you might also need *.md files as the version increment also generally updates changelog.md files.

**Review by weshaggard — APPROVED** [review]
Overall looks good. Lets get this in as soon as the test pipelines are green. I will hold my PR until after you merge and will handle merging the 2 together. 

I made a comment about the PackageVersionUpdate release stage which I think needs to be looked at a little closer. I suggest doing some test releases of the template package after you merges and have a close look at all the release steps and fix any issues independent of this PR.

File: eng/pipelines/templates/stages/archetype-java-release.yml
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-java-release.yml:278]: The package changelog gets checked out as part of the service directory path included in the `Paths` parameter.

I generally test template release with each functional commit, here's the most recent passing run: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=909741&view=logs&s=d0889692-cae7-51e0-3e1f-b1d8c62a4117&j=ea2ca241-627c-528c-676e-0ff517e2e7f6

**Pull request merged**
